src/remarx/sentence/corpus/tei_input.py

```diff
@@ -95,11 +106,39 @@ def get_page_footnotes(self) -> list[xmlmap.XmlObject]:
 
         return page_footnotes
 
+    def get_body_text_line_number(self, char_pos: int) -> int:
+        """
+        Return the TEI line number for the line preceding ``char_pos``.
+        """
+        if not hasattr(self, "_line_number_by_offset"):
+            self.get_body_text()
+
+        line_number_by_offset: OrderedDict[int, int] = getattr(
+            self, "_line_number_by_offset", OrderedDict()
+        )
+        offsets: list[int] = getattr(self, "_sorted_line_offsets", [])
+
+        # When there are no line breaks with line numbers, default to line 1
+        if not offsets:
+            return 1
+
+        line_number = line_number_by_offset[offsets[0]]
+        for offset in offsets:
+            if offset > char_pos:
+                break
+            line_number = line_number_by_offset[offset]
+        return line_number
+
     def get_body_text(self) -> str:
         """
         Extract body text content for this page, excluding footnotes and editorial content.
+        While collecting the text, build a mapping of character offsets to TEI line numbers.
         """
-        body_text_parts = []
+        body_text_parts: list[str] = []
+        line_number_by_offset: OrderedDict[int, int] = OrderedDict()
+        offsets: list[int] = []
+        char_offset = 0
```

Comment:
dict order by insertion order is guaranteed starting in python 3.7 and we're well past that, so no need to use ordered dict.

I also don't understand why you need to keep double track of the offsets. Please simplify to use the class instance variables throughout (instead of local and then copied to class instance), and just use the line number offset mapping.